File: src/main/scala/examples/ss4/Smallest.scala
package examples.ss4

object Smallest {
  
  def smallest0[A:Ordering](k: Int, pair: (List[A],List[A])):A = {
    def union(pair:(List[A], List[A])): List[A] = {
      pair match {
        case (xs, Nil) => xs
        case (Nil, ys) => ys
        case (x::xs, y::ys) => if (implicitly[Ordering[A]].lt(x,y)) x :: union (xs, y::ys)
                               else y :: union (x::xs, ys)
      }
    }
    union(pair)(k)
  }
  
  def smallest[A: Ordering](k: Int, pair: (List[A],List[A])):A = {
    pair match {
      case (Nil, ws) => ws(k)
      case (zs, Nil) => zs(k)
      case (zs, ws) =>
        val p = zs.length / 2
        val q = ws.length / 2
        val (xs,a::ys) = zs.splitAt(p)
        val (us,b::vs) = ws.splitAt(q)
        (implicitly[Ordering[A]].lt(a,b), k <= p + q) match {
          case (true,true)   => smallest(k, (zs,us))
          case (true,false)  => smallest(k-p-1, (ys,ws))
          case (false,true)  => smallest(k, (xs,ws))
          case (false,false) => smallest(k-q-1, (zs, vs))
        }
    }
  }
  
  def search[A: Ordering](pair: (Vector[A],Vector[A]), k:Int, pairX:(Int,Int), pairY:(Int,Int)): A = {
    val (xa,ya) = pair
    val (lx,rx) = pairX
    val (ly,ry) = pairY
    val mx = (lx + rx) / 2
    val my = (ly + ry) / 2
    if (lx == rx) ya(k + ly)
    else if (ly == ry) xa(k + lx)
    else (implicitly[Ordering[A]].lt(xa(mx), ya(my)), k <= mx - lx + my - ly) match {
      case (true, true)    => search(pair, k, (lx,rx), (ly,my))
      case (true, false)   => search(pair, k - (mx - lx) - 1, (mx + 1, rx), (ly, ry))
      case (false, true)   => search(pair, k, (lx, mx), (ly, ry))
      case (false , false) => search(pair, k - (my - ly) - 1, (lx,rx), (my + 1, ry))
    }
  }
  
  def smallest2[A: Ordering](k: Int, pair: (Vector[A],Vector[A])):A = {
    val (xa,ya) = pair
    search(pair, k, (0, xa.length), (0, ya.length))
  }
}

File: haskell/ss4/smallest.hs
import Data.Array

{-
smallest :: Ord a => Int -> ([a],[a]) -> a
smallest k (xs,ys) = union (xs,ys) !! k

union (xs,[]) = xs
union ([],ys) = ys
union (x:xs,y:ys) | x < y = x : union (xs, y:ys)
                  | x > y = y : union (x:xs, ys)
-}

-- |[a]| = length [a]
-- (xs ++ ys) !! k = if k < |xs| then xs !! k else ys !! (k - |xs|)

-- union (xs ++ ys, us ++ vs) = union (xs,us) ++ union (ys,vs)

smallest :: Ord a => Int -> ([a],[a]) -> a
smallest k ([],ws) = ws !! k
smallest k (zs,[]) = zs !! k
smallest k (zs,ws) =
  case (a < b,k <= p + q) of
    (True,True)   -> smallest k (zs,us)
    (True,False)  -> smallest (k-p-1) (ys,ws)
    (False,True)  -> smallest k (xs,ws)
    (False,False) -> smallest (k-q-1) (zs,vs)
  where p = (length zs) `div` 2
        q = (length ws) `div` 2
        (xs,a:ys) = splitAt p zs
        (us,b:vs) = splitAt q ws

smallest' :: Ord a => Int -> (Array Int a, Array Int a) -> a
smallest' k (xa,ya) = search k (0, m + 1) (0, n + 1)
  where
    (0,m) = bounds xa
    (0,n) = bounds ya
    search k (lx, rx) (ly, ry)
      | lx == rx = ya ! (k + ly)
      | ly == ry = xa ! (k + lx)
      | otherwise =
          case (xa ! mx < ya ! my, k <= mx - lx + my - ly) of
            (True,True)   -> search k (lx,rx) (ly,my)
            (True,False)  -> search (k - (mx - lx) - 1) (mx + 1, rx) (ly, ry)
            (False,True)  -> search k (lx, mx) (ly, ry)
            (False,False) -> search (k - (my - ly) - 1) (lx,rx) (my + 1, ry)
          where mx = (lx + rx) `div` 2
                my = (ly + ry) `div` 2



